include/proxy-wasm/context.h

```diff
@@ -0,0 +1,731 @@
+   * stream Context, this will be a Root Context id.
+   */
+  virtual void onCreate(uint32_t parent_context_id);
+
+  // Network
+
+  /**
+   * Call on a stream Context to indicate that a new network connection has been been created.
+   * Calls onStart().
+   */
+  virtual ProxyAction onNetworkNewConnection() {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+  /**
+   * Call on a stream Context to indicate that data has arrived from downstream (e.g. on the
+   * incoming port that was accepted and for which the proxy is acting as a server).
+   * @param data_length the amount of data in the buffer.
+   * @param end_of_stream is true if no more data will be arriving.
+   * @return indicates the subsequent behavior of this stream, e.g. continue proxying or pause
+   * proxying.
+   */
+  virtual ProxyAction onDownstreamData(int /* data_length */, bool /* end_of_stream */) {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+  /**
+   * Call on a stream Context to indicate that data has arrived from upstream (e.g. on the outgoing
+   * port that the proxy connected and for which the proxy is acting as a client).
+   * @param data_length the amount of data in the buffer.
+   * @param end_of_stream is true if no more data will be arriving.
+   * @return indicates the subsequent behavior of this stream, e.g. continue proxying or pause
+   * proxying.
+   */
+  virtual ProxyAction onUpstreamData(int /* data_length */, bool /* end_of_stream */) {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+
+  /*
+   * The source of the close event.
+   * Unknown is when the source is not known.
+   * Local is when the close was initiated by the proxy.
+   * Remote is when the close was recieved from the peer.
+   */
+  enum class CloseType : uint32_t {
+    Unknown = 0,
+    Local = 1,
+    Remote = 2,
+  };
+
+  /**
+   * Call on a stream context to indicate that the downstream connection has closed.
+   * @close_type is the source of the close.
+   */
+  virtual void onDownstreamConnectionClose(CloseType /* close_type */) { unimplemented(); }
+
+  /**
+   * Call on a stream context to indicate that the upstream connection has closed.
+   * @close_type is the source of the close.
+   */
+  virtual void onUpstreamConnectionClose(CloseType /* close_type */) { unimplemented(); }
+
+  // HTTP
+
+  /**
+   * Call on a stream context to indicate that the request headers have arrived.  Calls onCreate().
+   */
+  virtual ProxyAction onRequestHeaders() {
+    unimplemented();
+    return ProxyAction::Continue;
+  }
+
+  /**
+   * Call on a stream context to indicate that body data has arrived.
+   * @param body_buffer_length the amount of data in the body buffer.
+   * @param end_of_stream is true if no more body data will be arriving.
+   */
+  virtual ProxyAction onRequestBody(int /* body_buffer_length */, bool /* end_of_stream */) {
+    unimplemented();
+    return ProxyAction::Continue;
```

Comment:
Continue/Pause is not sufficient at least from an Envoy perspective. How are you planning on handling stop and buffer vs. stop and watermark?

Reply:
The SDK has not be updated yet.  @PiotrSikora is working on that.  When it is updated, the full ProxyAction will include those options.